### ISSUE TITLE
There's no way to make a controlled input in a node with vanilla React

### ISSUE THREAD

**Issue by AndrewRayCode** [wbkd/react-flow#2221]
### Describe the Bug

The react flow docs show an example of a node with an input in it. This only works because the input is uncontrolled using defaultValue. https://reactflow.dev/docs/examples/nodes/custom-node/

If you make the input controlled, every time you type in it, the cursor jumps to the end of the input. This is because the way the nodes are updated uses the **async** setState callback:

```
setNodes((nds) => ...
```

(If you update a controlled input's value async, the cursor jumps, it has to be set sync)

In the react flow example, the async state set to be done because the onChange handler is passed in node data in the example. Any data the onchange handler closes over in the parent component will be stale. Trying to update the onchange handler with new closure data on every keypress sounds confusing.

Mortiz made a controlled input with zustand, which works because the setter lives outside node data: https://discord.com/channels/771389069270712320/971342201701412915/983467015312396320

However there's no clear way to do this without zustand or a state management library.

This is half bug, half feature request. It would be great to have a first class way to handle controlled inputs in flow nodes.

One reason I want controlled inputs in nodes is to have a number input field along with a slider in the node, so that one changes the other. So I can't use `defaultValue` because both need to update in tandem

### Steps to Reproduce the Bug or Issue

Try to make a controlled input on a node

### Expected behavior

There to be a first class vanilla react way to do this

### Platform

- OS: MacOS
- Browser: Chrome
- Version: 10.2.1

**Comment by AndrewRayCode** [comment]
I think this should work, but it has the same cursor jumping issue: https://codesandbox.io/s/react-flow-renderer-cursor-jump-mc60w2?file=/src/App.js

To reproduce on this code sandbox:
- focus the text input in the node, and move the cursor to the beginning
- type in the input 
- the cursor jumps to the end of the input every time you type

The codesandbox implements this workaround I'm trying, where the onchange handler is passed down via context. https://codesandbox.io/s/context-onchange-controlled-input-ojg8bd This codesandbox doesn't have the cursor jumping issue, so I don't think it's an issue with passing the onChange in context

does rfr update nodes async when calling `setNodes`?

**Comment by moklick** [comment]
This is not an issue with React Flow. We are not doing node updates asyncally. The issue with your code is, that the `valSetter` is not memoized. If it's not memoized, the `valSetter` / `onChanges` updates the custom node before the actual input changes arrive.

Wrapping it with a `useCallback` fixes the issue: https://codesandbox.io/s/react-flow-renderer-cursor-jump-forked-ni12h4?file=/src/App.js:903-952

**Comment by AndrewRayCode** [comment]
Thank you very much for creating that demo. I was under the impression that the setState callback was an async operation, that impression looks to be incorrect.

This is strange, I'm doing basically the same thing in my real app, but I get the cursor jumping behavior. It's in a much more complicated setup so I wonder if some other async operation is at play here. Would not memoizing _other_ callbacks, like `onConnect`, cause this behavior if they changed every render in the parent?

Either way, I'll close this for now, thanks again for taking the time to give me updated code.

**Comment by moklick** [comment]
Sure! This is also super helpful for us to understand issues with the usage of the library better and might create more docs about these topics.

**Comment by AndrewRayCode** [comment]
Interestingly the only other callback i've found so far that causes the cursor to jump is that I was passing in an unmemoized `onNodeDoubleClick`. 

Is the best practice with rfr to memoize all callbacks fn props?

**Comment by moklick** [comment]
yep! I recently added this information to the docs https://reactflow.dev/docs/api/react-flow-props/#event-handlers

**Comment by nkosarek-zymergen** [comment]
Just a heads up for anyone else with the same issue: this can also happen if you have local state in your custom node that you update alongside the node's data field.

I forked @moklick's code sandbox with the `useCallback` fix, updated it to use React Flow v11.7.4, and added a local `numEdits` state variable in the custom node that gets incremented every time the color input is edited, and the cursor position reset behavior happens again:
https://codesandbox.io/s/react-flow-renderer-cursor-jump-forked-hx6snf?file=/src/ColorSelectNode.js

I assume this is happening for the same reason as the not-memoized callbacks example above - i.e. updates to the node data do not propagate immediately the next render but the local state update does.

**Comment by egorov-evrone** [comment]
@moklick hey) could you please take a look at the comment and one above?

I have somewhat similar issue. I use controlled nodes and edges. To control them there is a Context Provider that stores nodes and edges. There is also a custom node that makes use of the Context.

When I change data inside nodes array, the custom node first rerenders with obsolete data object. 

I forked your example and made some changes to illustrate the issue: https://codesandbox.io/p/sandbox/react-flow-renderer-cursor-jump-forked-p9vqhk?file=%2Fsrc%2FApp.js%3A49%2C1

After clicking on the "reset color" button, in the console we can see:
```
nodes, data [{…}] {color: '#1A192B'}
nodes, data [{…}] {color: '#CCCCCC'}
nodes, data [{…}] {color: '#CCCCCC'}
```

The first line shows that, on the first rerender, nodes array already have updated data. Whereas custom's node data object has the old color.

P. S. I assume, I can use [ReactFlowInstance](https://reactflow.dev/api-reference/types/react-flow-instance) to get nodes and update them inside custom nodes. However, current behaviour still seems to have a flaw. And normally, if I have some a state that data originates from, shouldn't I update exactly that one?

Considering, the comment above and this one, @moklick could you please take a look and tell if this is a bug or we doing something wrong?

Thanks!
